pkg/kv/kvserver/replica_destroy.go

```diff
@@ -186,8 +186,6 @@ func (r *Replica) destroyRaftMuLocked(ctx context.Context, nextReplicaID roachpb
 // is one, and removes the in-memory raft state.
 func (r *Replica) disconnectReplicationRaftMuLocked(ctx context.Context) {
 	r.raftMu.AssertHeld()
-	r.readOnlyCmdMu.Lock()
```

Comment:
This is because it's already held? Can we `r.readOnlyCmdMu.AssertHeld()` in that case (maybe on master only)?

Reply:
No, it's not held, but both callers have already called `destroyStatus.Set()` with it (temporarily) held. As such, the following code paths (including this one) are guaranteed to not conflict with read-only requests since they will either have completed already or will check `destroyStatus` and abort.

Reply:
The point about using `r.readOnlyCmdMu.AssertHeld()` to assert proper locking is still a good one. Did you give any thought to pulling accesses of `r.mu.destroyStatus` into methods on the replica and asserting proper handling of `r.mu` and `r.readOnlyCmdMu`?

> maybe on master only

That shouldn't be needed. `AssertHeld()`/`AssertRHeld()` should be no-ops on prod builds. The assertions are only armed in race builds.

Reply:
> Did you give any thought to pulling accesses of r.mu.destroyStatus into methods on the replica and asserting proper handling of `r.mu` and `r.readOnlyCmdMu`?

That's a good idea. I'm going to merge this and the other backports, so we have a fix ready to go, and then submit a follow-up PR to clean up the lock handling a bit. I'd also like to write a quick test for the write path in #46329 first, to see if it has a similar problem.

Reply:
#64459